Integrator spreadsheets

moritzcopter opened the pull request:
@mabrehm @SamLubbers  I have first basic working version and cleaned up the repo a bit.

The `update_integrator_spreadsheets.py` now pulls the most recent sign-ups from action network and pushes them to the Integration folder in [nextcloud](https://cloud.extinctionrebellion.nl/index.php/apps/files/?dir=/CloudXRNL/Circles/Integration_Home/Integration_Internal) for each local group.
If there is no spreadsheet for that group yet, it creates it, otherwise it appends new data while conserving e.g. formatting of the remaining sheet.

Main remaining issue I see is managing privacy: are we fine to share all names and email addresses with everyone who has access to that Nextcloud folder? Or shall we set up separate folders for each integrator to which only they have permission?
I am checking with John what is feasible in nextcloud, but let me know what you think.

martijnmartijnmartijnmartijn commented:
Awesome work Moritz. I think your second proposal (separate folders for each local group) seems good. Then, as you say, each integrator has access to only their data, which seems perfect. And it should be very doable Nextcloud-wise as far as I know.

I looked through the code, and it looks good to me. Didn't feel like spending lots of time commenting nitpicky shit, because we probably all have better things to do with our time :D

Some points on the functionality:
- Looking at the folder in Nextcloud, I only see like 6 local groups - is that correct? Or does your code somehow miss a few of them? Make sure you don't use the _"local_group"_ field that people have, as it's outdated. The local group can now be determined by the municipality (I think you also used this for the MM bot you made, so I think you have the distribution of municipalities over local groups, otherwise I have a spreadsheet somewhere I can share with an overview).
- I guess you still need to finalise placing all the data in the sheet, but make sure you don't forget to add the Action Network tags we use to keep people's interest. These all have names starting with "Interesse in x" where _"x"_ is some circle, joining actions, etc. 
- Also about these tags, it might be nice to create an extra tab in the spreadsheets where we place a (copy) of all the people with the _"Interesse in sporadisch helpen"_. These are people who wanna help out every now and then (with postering, calling people, etc). I imagine it's nice for local groups to have one clean list of people, so that they can easily reach out to them whenever they need some help.
- Do you think there might be ways in which someone might 'break' the sheet? I.e. moving columns, adding empty rows, etc. causing the loading of data to fail? If so, it might be useful to write some quick guidelines on how to the use sheets, and just add them at the top somewhere. Maybe we can all do some small tests trying to break it when we have a live version.

SamLubbers reviewed (changes requested):
@moritzcopter incredible that you got this done so quick. And even more incredible that I'm reviewing your code a few days after you make the pull request haha. Here is my feedback:

## File location (and who has access to it)

We'll probably continue this discussion with John on Nextcloud, but where I recommend storing the spreadsheet is within the folder that each local group has in Nextcloud. Specifically, I see it living here `XR Amsterdam > XR Amsterdam internal > Integrator > members.xlsx`. Only members of that local group have access to their internal folder. We need to see with John if we can restrict access of the integrator folder to only the integrator(s) of that local group. 

## Spreadsheet format

I see that the spreadsheet that you've created only have 3 columns. It would be useful if it had all the columns of the template spreadsheet, and to already set `Level of involvement = 1. Interested` and `Action = contact`. That way integrators can easily see and filter all those who they have to contact.

Also I see that you've made a duplicate of the template spreadsheet in the internal integration folder. To avoid duplicate data I suggest removing that one and using instead the one in the [integrator tools folder](https://cloud.extinctionrebellion.nl/index.php/f/377552)

@martijnmartijnmartijnmartijn I saw your comment about tags and other info stored in Action Network. Could you have a look at [the template](https://cloud.extinctionrebellion.nl/index.php/f/377552) and add the missing columns?

## Code

Unlike Martijn, I have some nitpicky comments :innocent: 

### Refactor reusable code

The code in `action_network.py` and `utils/members.py` would be useful to use in other projects as well. Could you move it to `rebel_management_utilities/api.py`? (maybe also rename `api.py` to `action_network.py`). The reason that inner package is also called `rebel_management_utilities` is so that we can package it and import it in other projects with `pip install git+https://github.com/xrnl/rebel_management_utilities`. With this package, the next time we want to test some action network script we could just import that package and use all those functions in a jupyter notebook :sunglasses: 

### Extract messages to markdown files

In the `post_signups_to_mattermost.py:15` I see that the message sent to integrators is in the python code. It would be great if you could move this to a markdown file so we can more easily update that message and see the results with a markdown reader.

I have also left some comments in the code. Besides this it looks really good! The modular code with clear function names makes it very easy to understand :muscle: :fire:

moritzcopter commented on local_group_support/update_integrator_spreadsheets.py:
```diff
@@ -0,0 +1,16 @@
+    start_date = datetime.date.today() - datetime.timedelta(days=3)
+    df = get_member_stats(start_date)
+
+    for local_group, df_grouped in df.groupby('local_group'):
```
`local_group` here is actually the one that is inferred from municipality

SamLubbers replied:
Oh good! in that case I have misread the code :sweat:

moritzcopter commented on local_group_support/utils/excel.py:
```diff
@@ -0,0 +1,32 @@
+        if startrow is None and sheet_name in writer.book.sheetnames:
+            startrow = writer.book[sheet_name].max_row
+
+        if truncate_sheet and sheet_name in writer.book.sheetnames:
+            idx = writer.book.sheetnames.index(sheet_name)
+            writer.book.remove(writer.book.worksheets[idx])
+            writer.book.create_sheet(sheet_name, idx)
```
I copied the function from stackoverflow 😇

SamLubbers replied:
haha cheeeeky :see_no_evil:

moritzcopter commented on local_group_support/update_integrator_spreadsheets.py:
```diff
@@ -0,0 +1,16 @@
+    write_to_spreadsheet
+
+if __name__ == "__main__":
+    start_date = datetime.date.today() - datetime.timedelta(days=3)
```
yeah that was for testing purposes

moritzcopter commented on local_group_support/utils/excel.py:
```diff
@@ -0,0 +1,32 @@
+from openpyxl import load_workbook
+
+
+def append_df_to_excel(filename, df, sheet_name='Sheet1', startrow=None,
```
Thats fair - fixed

moritzcopter commented on local_group_support/utils/mattermost.py:
```diff
@@ -0,0 +1,31 @@
+    headers = {'Authorization': get_mattermost_session_token()}
+    response = requests.post(url, headers=headers)
+    if response.status_code == 200:
+        return response.json()
```
Thanks for the comments guys - below some clarifications:

> * Looking at the folder in Nextcloud, I only see like 6 local groups - is that correct? Or does your code somehow miss a few of them? Make sure you don't use the _"local_group"_ field that people have, as it's outdated. The local group can now be determined by the municipality

It just takes all rebels available - I only tested it with rebels who signed up in the last few days, which only covered those 6 local groups. The local group here is inferred from the municipality.

> * Do you think there might be ways in which someone might 'break' the sheet? I.e. moving columns, adding empty rows, etc. causing the loading of data to fail? If so, it might be useful to write some quick guidelines on how to the use sheets, and just add them at the top somewhere.

Good point - when sharing the spreadsheet I will add some guidelines - Sams suggestion to use the first sheet by default should help already - otherwise the script doesnt really assume that certain columns are used in the sheet, so removing / adding any should not cause issues.

> * I guess you still need to finalise placing all the data in the sheet, but make sure you don't forget to add the Action Network tags we use to keep people's interest. These all have names starting with "Interesse in x" where _"x"_ is some circle, joining actions, etc.

I am looking into this - unfortunately action network does not return tags when querying rebels, so we need to additionally query the `tags` and `taggings` endpoints and map those to the members - custom fields are a bit more convenient in that regard

moritzcopter commented:
Thanks for your feedback guys - in the commits I pushed I:

- added exporting of tags
- refactored the code according to Sam's suggestions
- added logging messages to mattermost such that we have visibility if the script breaks once it is deployed

martijnmartijnmartijnmartijn commented:
Awesome stuff. Three more points:
1. Would be great if the form we send around after the weekly introduction sessions ([Dutch version](https://actionnetwork.org/forms/doe-mee-met-extinction-rebellion/) [English version](https://actionnetwork.org/forms/join-extinction-rebellion-3/)) could be mentioned as a specific source in the sheets, just like the website is now. Maybe call them "Introduction session".
2. I've added a new field to our forms where people can place general comments. These end up in the custom filed 'comments'. Could you add those the sheet as well?
3. I think it would be better if we write the labels on top of column in Dutch, as any integrator will have to speak Dutch anyway. You could call them (in the order they're in the sheet now) Naam, E-mail, Telefoon, Gemeente, Aangemeld via, Interesses, Volgende actie, Commentaar (for the comments).

martijnmartijnmartijnmartijn commented:
Oh and one more thing I think would be great to have:
4. For each AN form we have a Dutch and English version - would be great to add a column giving a person's language (or perhaps just list it in the source column if it's easier , i.e. "Website NL" instead of "Website")

SamLubbers commented on rebel_management_utilities/members.py:
```diff
@@ -101,21 +101,25 @@ def get_member_taggings(member):
+                 'languages_spoken': languages_spoken, 'email_address': email_address,
+                 'taggings': taggings,
+                 'phone_number': phone_number, **form} for form in forms]
+    except Exception as e:
```
Precisely which exceptions can occur here? Does it keep on working if a member doesn't have one of the custom fields?

moritzcopter replied:
Custom fields are not a problem - it breaks for example if members dont have a given_name field but yeah maybe a bit unnecessary to put it in a try / catch block, handle those cases explicitly now.

SamLubbers commented on local_group_support/update_integrator_spreadsheets.py:
```diff
@@ -20,7 +20,6 @@
-    except Exception as e:
-        post_to_channel(LOGGING_CHANNEL_ID, f'@all Failed to update integrator spreadsheet - {e}')
+            post_to_channel(LOGGING_CHANNEL_ID, f'Successfully updated integrator spreadsheet for {local_group}')
+        except Exception as e:
+            post_to_channel(LOGGING_CHANNEL_ID, f'@all Failed to update integrator spreadsheet for {local_group} - {e}')
```
Here it would also be good to know more precise information about the kind of exceptions that can occur

moritzcopter replied:
Im afraid quite a few things can go wrong here, e.g. connection / authentication / filenotfound errors + various issues related to handling excel in python that we might run into..

One case I already ran into with Armando's spreadsheet is that he tried to use the 'Tables' functionality in OnlyOffice - this doesnt seem to be supported by pyxl (package we use to read excel in python) which then cant read the sheet anymore.. Quite annoying as creating Tables using excel does not cause this issue, seems to be related to how OnlyOffice handles this - not sure there how to fix this, as its quite a corner case hard to find something on this

SamLubbers replied:
oh yeah those are some pretty obscure exceptions. In that case we can just leave it here and see what goes wrong :+1:

SamLubbers commented on local_group_support/utils/mattermost.py:
```diff
@@ -0,0 +1,25 @@
+from dotenv import load_dotenv
+
+BASE_URL = 'https://organise.earth/'
+LOGGING_CHANNEL_ID = 'gg3tdykut7f97xmg5kfpez7fuo'
```
which channel is this one? :thinking: It might be good to leave a comment next to each channel Id with the Mattermost team name and channel name.

moritzcopter replied:
done 👍

SamLubbers reviewed (commented):
Hey @moritzcopter it's so great that you've incorporated all of this feedback so quickly :raised_hands: I think that at this pace we'll be able to deploy this spreadsheet very soon.

Besides the few comments I left in the code, the only question I have left at the moment is if the new spreadsheets use the format that I suggested in my previous review (in the `spreadsheet format` section). If that is sorted and Martijn's comments are incorporated I think it's ready to go :tada:

SamLubbers commented:
and here is what I can add to the points @martijnmartijnmartijnmartijn 
>     3. I think it would be better if we write the labels on top of column in Dutch, as any integrator will have to speak Dutch anyway. 

you can make a clone of the English spreadsheet [in this folder](https://cloud.extinctionrebellion.nl/index.php/f/377551) and change the column names. We can then use that one by default.

> 4. For each AN form we have a Dutch and English version - would be great to add a column giving a person's language (or perhaps just list it in the source column if it's easier , i.e. "Website NL" instead of "Website")

Each person in action network has a field with the languages that they speak. These are automatically assigned depending on the language of the forms that they fill in. We could use that field to add the language to the spreadsheet.

> 5. Just noticed that only newly registered people are added.

Very very good point :clap: Another case that I foresee is if someone isn't active, but then is called by rebel ringers and now want to become active. Perhaps we should also have a tag or other mechanism to show that this person is recently interested. This aspect of adding newly interested people will definitely be something to refine once the first version of this script goes live.

moritzcopter commented:
Hey guys, thanks for your comments.

@martijnmartijnmartijnmartijn I added a comments columns and renamed the column names - below on the other points:

> 1. Would be great if the form we send around after the weekly introduction sessions ([Dutch version](https://actionnetwork.org/forms/doe-mee-met-extinction-rebellion/) [English version](https://actionnetwork.org/forms/join-extinction-rebellion-3/)) could be mentioned as a specific source in the sheets, just like the website is now. Maybe call them "Introduction session".
> 4. For each AN form we have a Dutch and English version - would be great to add a column giving a person's language (or perhaps just list it in the source column if it's easier , i.e. "Website NL" instead of "Website")

Thinking about those two points its actually easier if we export the name of the form directly instead of some mapped value - changed that.
Please make sure to set the "Administrative Title" field for new forms for that:
![image](https://user-images.githubusercontent.com/38844035/99151061-d37cfb80-2698-11eb-9c46-c90ff8af8e92.png)

> 5. Just noticed that only newly registered people are added. But I'm seeing a person who signed up months ago, and just now went to one of the introduction sessions on Sunday's, but these are also people we want to end up in the sheet. A solution would be to manually check new actions taken in the sign-ups of the forms given after the introduction sessions (NL/EN), and always adding these, if their original sign-up wasn't then.

We could filter by the date when those people were "modified" instead of "created" - downside of this would be that people who are already integrated but fill out a new form would appear in the sheets

martijnmartijnmartijnmartijn commented:
> We could filter by the date when those people were "modified" instead of "created" - downside of this would be that people who are already integrated but fill out a new form would appear in the sheets
> 
> Update: actually this returns too many "old" members - so thats not really a fix either..
I think the solution is to add new members _and_ everyone who fills in the form for the introduction sessions (so this can include 'old' people who have reregistered). This way, we get all new people, and only the old people who explicitly say that they want to get involved again. So we prevent a flooding of the sheets with old people.

moritzcopter commented:
> I think the solution is to add new members _and_ everyone who fills in the form for the introduction sessions (so this can include 'old' people who have reregistered). This way, we get all new people, and just the old that explicitly say that they want to get involved again. So we prevent a flooding of 'generic' old people.

Thats a good idea - pushed that change

martijnmartijnmartijnmartijn commented:
Nice! So I think there's just two things left now: the lay-out of the sheets and the overview of local groups.

I saw the the overview of local groups that you use was somewhat outdated so I updated it. Could you update "local_group_support/config/config.json" with this version: https://cloud.extinctionrebellion.nl/index.php/f/385530   I added new local groups, and updated local groups that have changed. 

For the lay-out of the sheets, I merged the lay-out that you used, Moritz, with the one Sam proposed: https://cloud.extinctionrebellion.nl/index.php/f/385524    I think this version will do. I also added some comments on top for integrators on how to use it. @SamLubbers could you check whether this still works for your guide. And @moritzcopter could you use this lay-out for all the sheets (after Sam gives his okay). If the comments on top of the sheet make the implementation harder, also feel free to remove them.

moritzcopter commented:
Thanks for updating the config Martijn, I pushed the new version.

Regarding layout I like that you included some instructions and the clear separations between columns that are updated and "manual" columns. I think to keep this its easiest to just create the sheets manually (I can do that once its ready).
Small change: I added the "Gemeente" column as it could be useful in the "rural" areas of the Netherlands ;)

Could you add to the instructions that they cant use the "Table" feature (Insert > Table) as python cant read the sheets anymore otherwise?
If you want to be precise, they actually can use it, but they cant have the Table extend over "infinite" rows by only selecting columns when creating it, but instead have to create a finite "box".
I tried a lot of things / libraries to fix this, but didnt manage to unfortunately.

martijnmartijnmartijnmartijn commented:
Good one to add the Gemeente! I've just added a note about the tables feature - I just told them to avoid it in general. 

I think that's it right? Sam and I we're hoping to contact all the integrators on Wednesday - could you copy the sheets, and run the script to fill them all up by then? Then we can start sending it around. I think Sam is gonna do one last code review as well, so after that we should be fine to merge this :)

moritzcopter commented:
> could you copy the sheets, and run the script to fill them all up by then?

I think so - is the layout ready or will you still change something?

martijnmartijnmartijnmartijn commented:
> Sure - is the layout ready or will you still change something?

Nope, all done. I added the right links in there just now.

moritzcopter commented:
Done - I also pulled the user instructions out into a readme file, this way we can update them more easily going forward and they are nicely displayed in the folder view in nextcloud.

martijnmartijnmartijnmartijn commented:
Looking at some of the sheets for smaller groups, I notice that they don't really have that many new rebels signing up. It think for those groups it would be nice if they could somehow request more rebels from in the past, so that they can call them as well if they have the time. Maybe an idea could be to have a cell at the top of the sheet which specifies the date from which on people are added. So if you put 15/07/2020 you get all rebels who signed up (or filled in the introduction session sheet) after that date. That way they could manually select what data they want, and have their sheet update at the next synchronisation. Would this be a lot of work to implement? Not necessarily right now of course, could also be something we do later on after people have started using it

SamLubbers commented:
> For the lay-out of the sheets, I merged the lay-out that you used, Moritz, with the one Sam proposed

Yes this layout is fantastic! But now I see three almost identical files in Nextcloud. @martijnmartijnmartijnmartijn could you delete the ones we are not using?

SamLubbers reviewed (approved):
@moritzcopter thanks for all the changes in this past week(s) :raised_hands: :100: I went through all the new commits and it all looks good! @martijnmartijnmartijnmartijn thanks again for updating the spreadsheet and for the quick feedback.

I'm approving and merging this PR so we can start using it to update the spreadsheets. Later we can add more and smaller improvements, such as showing more people for smaller groups.

SamLubbers commented:
> Looking at some of the sheets for smaller groups, I notice that they don't really have that many new rebels signing up. It think for those groups it would be nice if they could somehow request more rebels from in the past.

@martijnmartijnmartijnmartijn good observation! Could you [make a new issue](https://github.com/xrnl/rebel_management_utilities/issues/new/choose) for this? Then we can continue the discussion there and implement it when we have time. 

As some preliminary thought, I think we can keep the application simpler and more reliable if we don't have any inputs on the spreadsheet that alter the behaviour of the application. The way that we could overcome this automatically is being counting the number of people in a local group, and if it's lower than X people per X days we can add the contact details of more people who signed up earlier.

moritzcopter commented:
Thanks for reviewing Sam!

@martijnmartijnmartijnmartijn I agree with Sam that having inputs from the spreadsheet will make things a bit complicated. Since the low number of people is mainly a one-time issue since we will include new sign-ups either way, I will just re-run the script locally with an earlier ´start_date´ and limit the total amount of people that we push to 50 per group such that we dont clutter the sheets of the larger groups

SamLubbers commented:
@moritzcopter I think this approach could work now, but what if at some point there are more than 50 sign up in one week for one local group? In such cases you actually would want to add them. So the ideal solution always adds recent signups, sign-ups from people that went to the intro event, and if there are less than X sign-ups in a local group, add older sign ups.

moritzcopter commented:
Yes, thats how it works right now 😉
